File: app/src/main/java/com/supercilex/robotscouter/ui/scout/viewholder/template/SpinnerTemplateViewHolder.java
package com.supercilex.robotscouter.ui.scout.viewholder.template;

import android.view.View;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;

import com.supercilex.robotscouter.R;
import com.supercilex.robotscouter.data.model.metrics.ListMetric;
import com.supercilex.robotscouter.ui.scout.template.SpinnerTemplateDialog;
import com.supercilex.robotscouter.ui.scout.viewholder.SpinnerViewHolder;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.Map;

import static com.supercilex.robotscouter.util.ConstantsKt.FIREBASE_VALUE;

public class SpinnerTemplateViewHolder extends SpinnerViewHolder implements ScoutTemplateViewHolder {
    public SpinnerTemplateViewHolder(View itemView) {
        super(itemView);
    }

    @Override
    public void bind() {
        super.bind();
        getName().setOnFocusChangeListener(this);
    }

    @Override
    protected ArrayAdapter<String> getAdapter(ListMetric listMetric) {
        Map<String, String> items = new LinkedHashMap<>();
        items.put(getMetric().getRef().push().getKey(),
                  itemView.getContext().getString(R.string.edit_spinner_items));
        items.putAll(listMetric.getValue());
        return new ArrayAdapter<>(itemView.getContext(),
                                  android.R.layout.simple_spinner_item,
                                  new ArrayList<>(items.values()));
    }

    @Override
    public void onItemSelected(AdapterView parent, View view, int itemPosition, long id) {
        if (itemPosition == 0) {
            disableAnimations();
            updateMetricName(getName().getText().toString());

            ListMetric listMetric = (ListMetric) getMetric();
            SpinnerTemplateDialog.Companion.show(getManager(),
                                                 getMetric().getRef().child(FIREBASE_VALUE),
                                                 listMetric.getSelectedValueKey());
            mSpinner.setSelection(indexOfKey(listMetric.getSelectedValueKey()));
        } else {
            super.onItemSelected(parent, view, itemPosition - 1, id);
        }
    }

    @Override
    protected int indexOfKey(String key) {
        return super.indexOfKey(key) + 1;
    }

    @Override
    public void requestFocus() {
        getName().requestFocus();
    }

    @Override
    public void onFocusChange(View v, boolean hasFocus) {
        if (!hasFocus) updateMetricName(getName().getText().toString());
    }
}

File: app/src/main/java/com/supercilex/robotscouter/ui/scouting/template/viewholder/SpinnerTemplateViewHolder.kt
package com.supercilex.robotscouter.ui.scouting.template.viewholder

import android.view.View
import android.widget.AdapterView
import android.widget.ArrayAdapter
import com.supercilex.robotscouter.R
import com.supercilex.robotscouter.data.model.Metric
import com.supercilex.robotscouter.ui.scouting.scout.viewholder.SpinnerViewHolder
import com.supercilex.robotscouter.ui.scouting.template.SpinnerTemplateDialog
import com.supercilex.robotscouter.util.FIREBASE_VALUE
import java.util.ArrayList
import java.util.LinkedHashMap

class SpinnerTemplateViewHolder(itemView: View) : SpinnerViewHolder(itemView), TemplateViewHolder {
    override fun bind() {
        super.bind()
        name.onFocusChangeListener = this
    }

    override fun getAdapter(listMetric: Metric.List): ArrayAdapter<String> {
        val items = LinkedHashMap<String, String>()
        items.put(metric.ref.push().key, itemView.context.getString(R.string.edit_spinner_items))
        items.putAll(listMetric.value)
        return ArrayAdapter(
                itemView.context, android.R.layout.simple_spinner_item, ArrayList(items.values))
    }

    override fun onItemSelected(parent: AdapterView<*>, view: View, itemPosition: Int, id: Long) {
        if (itemPosition == 0) {
            disableAnimations()
            updateMetricName(name.text.toString())

            SpinnerTemplateDialog.show(
                    manager, metric.ref.child(FIREBASE_VALUE), metric.selectedValueKey)
            spinner.setSelection(indexOfKey(metric.selectedValueKey))
        } else {
            super.onItemSelected(parent, view, itemPosition - 1, id)
        }
    }

    override fun indexOfKey(key: String?) = super.indexOfKey(key) + 1

    override fun requestFocus() {
        name.requestFocus()
    }

    override fun onFocusChange(v: View, hasFocus: Boolean) {
        if (!hasFocus) updateMetricName(name.text.toString())
    }
}
